extensions/notebook/src/jupyter/jupyterServerInstallation.ts

```diff
@@ -356,6 +406,92 @@ export class JupyterServerInstallation {
 		}
 	}
 
+	/**
+	 * Prompts user to upgrade certain python packages if they're below the minimum expected version.
+	 */
+	public promptForPackageUpgrade(): Promise<void> {
+		if (this._usingConda) {
+			return this.upgradeCondaPackages(true);
+		} else {
+			return this.upgradePythonPackages(true);
+		}
+	}
+
+	private async upgradePythonPackages(promptForInstall: boolean): Promise<void> {
+		let installedPackages = await this.getInstalledPipPackages();
+		let pkgVersionMap = new Map<string, string>();
+		installedPackages.forEach(pkg => pkgVersionMap.set(pkg.name, pkg.version));
+
+		let packagesToInstall: PythonPkgDetails[] = [];
+		this._expectedPythonPackages.forEach(expectedPkg => {
+			let installedPkgVersion = pkgVersionMap.get(expectedPkg.name);
+			if (!installedPkgVersion || utils.comparePackageVersions(installedPkgVersion, expectedPkg.version) < 0) {
+				packagesToInstall.push(expectedPkg);
+			}
+		});
+
+		if (packagesToInstall.length > 0) {
+			let doUpgrade: boolean;
+			if (promptForInstall) {
+				doUpgrade = await this._prompter.promptSingle<boolean>(<IQuestion>{
+					type: QuestionTypes.confirm,
+					message: localize('confirmPipUpgrade', 'Some installed pip packages need to be upgraded. Would you like to upgrade them now?'),
+					default: true
+				});
+			} else {
+				doUpgrade = true;
+			}
+			if (doUpgrade) {
+				await this.installPipPackages(packagesToInstall);
+			}
+		}
+	}
+
+	private async upgradeCondaPackages(promptForInstall: boolean): Promise<void> {
```

Comment:
This method is quite similar to the method above in structure. Can we consolidate them somehow?

Reply:
Maybe call the pip-related package checking from a common method?

Reply:
Originally I had them as a single method, but all the branching if cases made it hard to read. Given that there are only these two install methods, I'm fine leaving these as is for now.